File: Tests/TokenSpendieTests/CodexCredentialsStoreTests.swift
import XCTest
@testable import TokenSpendie

final class CodexCredentialsStoreTests: XCTestCase {
    private var dir: URL!

    override func setUpWithError() throws {
        dir = FileManager.default.temporaryDirectory
            .appendingPathComponent("codex-store-\(UUID().uuidString)", isDirectory: true)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
    }

    override func tearDownWithError() throws {
        try? FileManager.default.removeItem(at: dir)
    }

    private func write(_ json: String) throws -> CodexCredentialsStore {
        let url = dir.appendingPathComponent("auth.json")
        try Data(json.utf8).write(to: url)
        return CodexCredentialsStore(fileURL: url)
    }

    private static let fullAuthJSON = #"""
    {
      "OPENAI_API_KEY": null,
      "tokens": {
        "id_token": "id.jwt",
        "access_token": "access.jwt",
        "refresh_token": "refresh-1",
        "account_id": "account-123"
      },
      "last_refresh": "2026-06-01T00:00:00Z",
      "future_field": {"keep": true}
    }
    """#

    func testLoadParsesTokens() throws {
        let store = try write(Self.fullAuthJSON)
        let creds = try store.load()
        XCTAssertEqual(creds.accessToken, "access.jwt")
        XCTAssertEqual(creds.refreshToken, "refresh-1")
        XCTAssertEqual(creds.accountId, "account-123")
        XCTAssertEqual(creds.lastRefresh,
                       UsageDecoder.parseDate("2026-06-01T00:00:00Z"))
    }

    func testDetectTrueOnlyWithNonEmptyAccessToken() throws {
        XCTAssertTrue(try write(Self.fullAuthJSON).detect())
        // API-key-only file → not detected (spec: same policy as Gemini API-key).
        XCTAssertFalse(try write(#"{"OPENAI_API_KEY": "sk-x"}"#).detect())
        XCTAssertFalse(try write(#"{"tokens": {"access_token": ""}}"#).detect())
        let missing = CodexCredentialsStore(
            fileURL: dir.appendingPathComponent("nope.json"))
        XCTAssertFalse(missing.detect())
    }

    func testLoadMissingOrMalformedThrowsReauthRequired() throws {
        let missing = CodexCredentialsStore(
            fileURL: dir.appendingPathComponent("nope.json"))
        XCTAssertThrowsError(try missing.load()) {
            XCTAssertEqual($0 as? ProviderError, .reauthRequired)
        }
        let malformed = try write("not json")
        XCTAssertThrowsError(try malformed.load()) {
            XCTAssertEqual($0 as? ProviderError, .reauthRequired)
        }
    }

    func testNeedsRefreshAfterEightDays() throws {
        let creds = try write(Self.fullAuthJSON).load()
        let lastRefresh = creds.lastRefresh!
        XCTAssertFalse(creds.needsRefresh(now: lastRefresh.addingTimeInterval(7 * 86400)))
        XCTAssertTrue(creds.needsRefresh(now: lastRefresh.addingTimeInterval(9 * 86400)))
        // Exactly 8 days is NOT "older than 8 days" — strict >.
        XCTAssertFalse(creds.needsRefresh(now: lastRefresh.addingTimeInterval(8 * 86400)))
        let noStamp = CodexCredentials(accessToken: "a", refreshToken: "r",
                                       idToken: nil, accountId: nil, lastRefresh: nil)
        XCTAssertTrue(noStamp.needsRefresh(now: Date()))
    }

    func testSavePreservesUnknownFieldsAndUpdatesTokens() throws {
        let store = try write(Self.fullAuthJSON)
        let updated = CodexCredentials(accessToken: "access-2", refreshToken: "refresh-2",
                                       idToken: "id-2", accountId: "account-123",
                                       lastRefresh: nil)
        try store.save(updated, refreshedAt: Date(timeIntervalSince1970: 1_780_000_000))

        let reloaded = try store.load()
        XCTAssertEqual(reloaded.accessToken, "access-2")
        XCTAssertEqual(reloaded.refreshToken, "refresh-2")
        XCTAssertEqual(reloaded.idToken, "id-2")
        XCTAssertNotNil(reloaded.lastRefresh)

        let raw = try JSONSerialization.jsonObject(
            with: Data(contentsOf: store.fileURL)) as! [String: Any]
        XCTAssertNotNil(raw["future_field"], "unknown top-level fields must survive")
        XCTAssertTrue(raw["OPENAI_API_KEY"] is NSNull, "null API key must survive")
    }

    func testSaveOnMissingFileCreatesOwnerOnlyPermissions() throws {
        let store = CodexCredentialsStore(fileURL: dir.appendingPathComponent("auth.json"))
        let creds = CodexCredentials(accessToken: "a", refreshToken: "r",
                                     idToken: nil, accountId: nil, lastRefresh: nil)
        try store.save(creds, refreshedAt: Date())
        let attrs = try FileManager.default.attributesOfItem(atPath: store.fileURL.path)
        XCTAssertEqual((attrs[.posixPermissions] as? NSNumber)?.int16Value, 0o600)
    }

    func testDefaultURLHonorsCodexHome() {
        let url = CodexCredentialsStore.defaultURL(
            environment: ["CODEX_HOME": "/tmp/custom-codex"],
            home: URL(fileURLWithPath: "/Users/x"))
        XCTAssertEqual(url.path, "/tmp/custom-codex/auth.json")
        let fallback = CodexCredentialsStore.defaultURL(
            environment: [:], home: URL(fileURLWithPath: "/Users/x"))
        XCTAssertEqual(fallback.path, "/Users/x/.codex/auth.json")
    }
}

File: windows/tests/TokenSpendie.Windows.Tests/Data/CodexCredentialsStoreTests.cs
using System;
using System.IO;
using System.Text.Json;
using FluentAssertions;
using TokenSpendie.Windows.Data;
using TokenSpendie.Windows.Models;
using Xunit;

namespace TokenSpendie.Windows.Tests.Data;

public sealed class CodexCredentialsStoreTests : IDisposable
{
    private readonly string _dir =
        Path.Combine(Path.GetTempPath(), "codex-store-" + Guid.NewGuid().ToString("N"));

    public CodexCredentialsStoreTests() => Directory.CreateDirectory(_dir);
    public void Dispose() { try { Directory.Delete(_dir, true); } catch { } }

    private const string FullAuthJson = """
    {
      "OPENAI_API_KEY": null,
      "tokens": {
        "id_token": "id.jwt",
        "access_token": "access.jwt",
        "refresh_token": "refresh-1",
        "account_id": "account-123"
      },
      "last_refresh": "2026-06-01T00:00:00Z",
      "future_field": {"keep": true}
    }
    """;

    private CodexCredentialsStore Write(string json)
    {
        var path = Path.Combine(_dir, "auth.json");
        File.WriteAllText(path, json);
        return new CodexCredentialsStore(path);
    }

    [Fact]
    public void LoadParsesTokens()
    {
        var creds = Write(FullAuthJson).Load();
        creds.AccessToken.Should().Be("access.jwt");
        creds.RefreshToken.Should().Be("refresh-1");
        creds.AccountId.Should().Be("account-123");
        creds.LastRefresh.Should().Be(DateTimeOffset.Parse("2026-06-01T00:00:00Z"));
    }

    [Fact]
    public void DetectTrueOnlyWithNonEmptyAccessToken()
    {
        Write(FullAuthJson).Detect().Should().BeTrue();
        Write("""{"OPENAI_API_KEY": "sk-x"}""").Detect().Should().BeFalse();
        Write("""{"tokens": {"access_token": ""}}""").Detect().Should().BeFalse();
        new CodexCredentialsStore(Path.Combine(_dir, "nope.json")).Detect().Should().BeFalse();
    }

    [Fact]
    public void LoadMissingOrMalformedThrowsReauthRequired()
    {
        var missing = new CodexCredentialsStore(Path.Combine(_dir, "nope.json"));
        missing.Invoking(s => s.Load()).Should().Throw<ProviderReauthRequiredException>();
        Write("not json").Invoking(s => s.Load()).Should().Throw<ProviderReauthRequiredException>();
    }

    [Fact]
    public void NeedsRefreshAfterEightDays()
    {
        var creds = Write(FullAuthJson).Load();
        var stamp = creds.LastRefresh!.Value;
        creds.NeedsRefresh(stamp.AddDays(7)).Should().BeFalse();
        creds.NeedsRefresh(stamp.AddDays(8)).Should().BeFalse("exactly 8 days is not OLDER than 8 days");
        creds.NeedsRefresh(stamp.AddDays(9)).Should().BeTrue();
        (creds with { LastRefresh = null }).NeedsRefresh(DateTimeOffset.UtcNow).Should().BeTrue();
    }

    [Fact]
    public void SavePreservesUnknownFieldsAndUpdatesTokens()
    {
        var store = Write(FullAuthJson);
        var updated = new CodexCredentials("access-2", "refresh-2", "id-2", "account-123", null);
        store.Save(updated, DateTimeOffset.FromUnixTimeSeconds(1_780_000_000));

        var reloaded = store.Load();
        reloaded.AccessToken.Should().Be("access-2");
        reloaded.RefreshToken.Should().Be("refresh-2");
        reloaded.LastRefresh.Should().NotBeNull();

        using var doc = JsonDocument.Parse(File.ReadAllText(store.FilePath));
        doc.RootElement.TryGetProperty("future_field", out _)
            .Should().BeTrue("unknown top-level fields must survive");
        doc.RootElement.GetProperty("OPENAI_API_KEY").ValueKind
            .Should().Be(JsonValueKind.Null);
    }

    [Fact]
    public void DefaultPathHonorsCodexHome()
    {
        CodexCredentialsStore.DefaultPath(codexHome: @"C:\custom", userProfile: @"C:\Users\x")
            .Should().Be(Path.Combine(@"C:\custom", "auth.json"));
        CodexCredentialsStore.DefaultPath(codexHome: null, userProfile: @"C:\Users\x")
            .Should().Be(Path.Combine(@"C:\Users\x", ".codex", "auth.json"));
    }
}
